### PR TITLE
Add options for controlling emission of randomization code.

### LINKED ISSUE #4202: [FIRRTL] Feature Request: Please support options for omission of the randomization code
Scala FIRRTL Compiler supported annotations to control the emission of randomization code at the top of the file, which can be quite noisy. In particular the `CustomDefaultRegisterEmission.disableRandomization` flag and ` case "disableMemRandomization" => CustomDefaultMemoryEmission(MemoryNoInit)` was useful when trying to emit instructive examples and documentation where the initialization code is huge and not exactly educational.

See https://github.com/chipsalliance/firrtl/blob/b027eb466b033a0a9d229f19feb931ddb292a9fa/src/main/scala/firrtl/EmissionOption.scala#L15

other discussion here: https://discourse.llvm.org/t/how-to-disable-randomize-in-firtool-generated-sv/65990

### PR DESCRIPTION
**Pull request by dtzSiFive**
Primarily introduce `firtool` options `--disable-{reg,mem,all}-randomization`.

Fixes #4202 .

### CODE REVIEW

File: lib/Dialect/SV/Transforms/HWMemSimImpl.cpp [hunk 1]
```diff
@@ -417,135 +424,149 @@ void HWMemSimImpl::generateMemory(HWModuleOp op, FirMemory mem) {
...
-    b.create<sv::IfDefOp>("RANDOMIZE_MEM_INIT", [&]() {
-      initvar = moduleNamespace.newName("initvar");
-      b.create<sv::VerbatimOp>("integer " + Twine(initvar) + ";\n");
-      randomMemReg = b.create<sv::RegOp>(
-          b.getIntegerType(llvm::divideCeil(mem.dataWidth, randomWidth) *
-                           randomWidth),
-          b.getStringAttr("_RANDOM_MEM"),
-          b.getStringAttr(moduleNamespace.newName("_RANDOM_MEM")));
-    });
+  if (!disableMemRandomization || !disableRegRandomization) {
```
Comment (seldridge) [inline lib/Dialect/SV/Transforms/HWMemSimImpl.cpp:427]: Can the PR decrease the nesting by inverting the condition and early exiting or by refactoring the randomization logic into a method?

File: test/Dialect/SV/hw-memsim.mlir [hunk 2]
```diff
@@ -1,10 +1,14 @@
...
+// RUN: circt-opt -pass-pipeline="hw-memory-sim{strip-mux-pragmas}" %s | FileCheck %s --check-prefix=STRIP --implicit-check-not sv.attributes
+// RUN: circt-opt -pass-pipeline="hw-memory-sim{disable-mem-randomization}" %s | FileCheck %s --check-prefix=DISABLE_RANDOM --implicit-check-not RANDOMIZE_MEM
+// RUN: circt-opt -pass-pipeline="hw-memory-sim{disable-reg-randomization}" %s | FileCheck %s --check-prefix=DISABLE_RANDOM --implicit-check-not RANDOMIZE_REG
+// RUN: circt-opt -pass-pipeline="hw-memory-sim{disable-mem-randomization disable-reg-randomization}" %s | FileCheck %s --check-prefix=DISABLE_RANDOM --implicit-check-not RANDOMIZE_REG --implicit-check-not RANDOMIZE_MEM
 
 hw.generator.schema @FIRRTLMem, "FIRRTL_Memory", ["depth", "numReadPorts", "numWritePorts", "numReadWritePorts", "readLatency", "writeLatency", "width", "readUnderWrite", "writeUnderWrite", "writeClockIDs"]
 
 //CHECK-LABEL: @complex
+// STRIP-LABEL: @complex
+// DISABLE_RANDOM-LABEL: @complex
```
Comment (seldridge) [inline test/Dialect/SV/hw-memsim.mlir:11]: The PR could consider using check-prefixes to share a common prefix for label checking.

File: tools/firtool/firtool.cpp [hunk 3]
```diff
@@ -322,6 +322,22 @@ static cl::opt<bool> etcDisableModuleInlining(
...
+enum class RandomKind { None, Mem, Reg, All };
+
+static cl::opt<RandomKind> disableRandom(
+    cl::desc("Disable random initialization code (may break semantics!)"),
+    cl::values(clEnumValN(RandomKind::Mem, "disable-mem-randomization",
+                          "Disable emission of memory randomization code"),
+               clEnumValN(RandomKind::Reg, "disable-reg-randomization",
+                          "Disable emission of register randomization code"),
+               clEnumValN(RandomKind::All, "disable-all-randomization",
+                          "Disable emission of register randomization code")),
```
Comment (seldridge) [inline tools/firtool/firtool.cpp:334]: This should indicate it disables all randomization.

**Review by seldridge — APPROVED** [review]
Nice. This is a big help for readability of examples.

Can this PR also turn off any stable randomization stuff?

File: lib/Dialect/SV/Transforms/HWMemSimImpl.cpp [hunk 4]
```diff
@@ -417,135 +424,149 @@ void HWMemSimImpl::generateMemory(HWModuleOp op, FirMemory mem) {
...
-        auto repetitionCount = llvm::divideCeil(mem.dataWidth, randomWidth);
-        if (repetitionCount > 1)
-          s << "{";
-        for (size_t i = 0; i != repetitionCount; ++i) {
-          if (i > 0)
-            s << ", ";
-          s << "{`RANDOM}";
+      b.create<sv::InitialOp>([&]() {
+        b.create<sv::VerbatimOp>("`INIT_RANDOM_PROLOG_");
+
```
Comment (uenoku) [inline lib/Dialect/SV/Transforms/HWMemSimImpl.cpp:466]: Possibly we can remove initial block and verbatim `INIT_RANDOM_PROLOG`?

File: lib/Dialect/SV/Transforms/HWMemSimImpl.cpp (hunk 1 above)
Comment (dtzSiFive) [reply lib/Dialect/SV/Transforms/HWMemSimImpl.cpp]: Good call, looks like code after the conditional can be safely hoisted before, so can move to early-exit.  Thanks!

File: test/Dialect/SV/hw-memsim.mlir (hunk 2 above)
Comment (dtzSiFive) [reply test/Dialect/SV/hw-memsim.mlir]: Nice! Done!

File: tools/firtool/firtool.cpp (hunk 3 above)
Comment (dtzSiFive) [reply tools/firtool/firtool.cpp]: Whoops, fixed.  Thanks!

File: lib/Dialect/SV/Transforms/HWMemSimImpl.cpp (hunk 4 above)
Comment (dtzSiFive) [reply lib/Dialect/SV/Transforms/HWMemSimImpl.cpp]: We will omit them if both mem and reg randomization is disabled, e.g., with `--disable-all-randomization`.  I'd expect that's the main option that'll be used (of the new options).

**Comment by dtzSiFive** [comment]
> Can this PR also turn off any stable randomization stuff?

Specifying `--disable-reg-randomization` turns off `firrtl-randomize-register-init` (in `firtool.cpp`), which I think accomplishes what you're asking, is that what you mean?

**Review by mikeurbach — COMMENTED** [review]
I'm going to be AFK today for travel, so I won't be able to look deeply at this, but I think it makes sense in general.

File: lib/Dialect/SV/Transforms/HWMemSimImpl.cpp (hunk 4 above)
Comment (uenoku) [reply lib/Dialect/SV/Transforms/HWMemSimImpl.cpp]: Make sense, I was missing early exit. Thanks!

**Comment by dtzSiFive** [comment]
Thanks folks!  Shouldn't have impact unless enabled, and if we want it to turn off more than it does we can improve it in the future, and this gets us nicer examples in this release :+1:.

**Pull request merged**